Compute code-coverage of backend files

bajtos opened the pull request:
When the module is running under a code-coverage tool nyc (tap --cov), then load instrumented backend files from "backend.cov" and write code coverage data to a file whenever the debugger backend is stopped.

Connect to strongloop-internal/scrum-loopback#469

/to @rmg @bnoordhuis please review

I am not entirely happy that we will be shipping code-coverage related hacks to npmjs, but then I don't know any better way how to compute code coverage of backend files running in a custom non-Node isolate. Do you?

Current code coverage report:

```
-------------------------------------------|----------|----------|----------|----------|----------------|
File                                       |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
-------------------------------------------|----------|----------|----------|----------|----------------|
 Users/bajtos/src/strong/debugger/backend/ |    79.37 |    64.01 |    91.87 |    83.03 |                |
  context.js                               |    78.42 |    62.82 |    95.45 |    79.39 |... 311,312,332 |
  convert.js                               |    76.92 |    64.29 |    95.83 |    78.07 |... 280,281,283 |
  debugger-agent.js                        |    79.84 |     58.9 |    91.67 |     88.6 |... 300,301,302 |
  page-agent.js                            |    69.23 |       60 |    63.64 |       75 |... 62,84,88,92 |
  runtime-agent.js                         |    89.06 |       80 |      100 |    93.22 | 58,122,178,185 |
 __root__/                                 |       88 |       70 |     87.5 |     91.3 |                |
  ./index.js                               |       88 |       70 |     87.5 |     91.3 |    34,35,82,83 |
-------------------------------------------|----------|----------|----------|----------|----------------|
All files                                  |    80.19 |     64.4 |     91.6 |    83.81 |                |
-------------------------------------------|----------|----------|----------|----------|----------------|
```

Commit:
Compute code-coverage of backend files

When the module is running under a code-coverage tool nyc (tap --cov),
then load instrumented backend files from "backend.cov" and write
code coverage data to a file whenever the debugger backend is stopped.

bajtos commented on package.json:
```diff
@@ -16,7 +16,8 @@
   "scripts": {
     "install": "node-gyp rebuild",
-    "pretest": "jscs . && jshint .",
+    "instrument": "node node_modules/tap/node_modules/nyc/node_modules/istanbul/lib/cli.js instrument --output backend.cov --no-compact --preserve-comments backend",
```
This is rather ugly, but I'd rather have a build that fails because the dependency tree changed, than to have a subtly broken code coverage report caused by running a different istanbul version for instrumentation and report generation.

And yes, I do realise I'll have to rework this anyway after we add a `npm@3` CI slave.

rmg commented on package.json:
```diff
@@ -16,7 +16,8 @@
-    "pretest": "jscs . && jshint .",
+    "instrument": "node node_modules/tap/node_modules/nyc/node_modules/istanbul/lib/cli.js instrument --output backend.cov --no-compact --preserve-comments backend",
+    "pretest": "jscs . && jshint . && npm run instrument",
     "test": "tap test/lab/*.test.js test/*.test.js"
```
To actually generate coverage on CI, either the `test` script needs to be modified to include `--coverage --coverage-report=cobertura` or build-framework needs to be modified it inject it when `tap@1.x` is detected.

bajtos replied:
Since I am going to change build-framework anyways, I prefer to inject these two extra args there.

rmg commented:
Reading through the implementation, I don't think it's _quite_ as bad as you described. Or at least, you've done it the cleanest way I can think of.

Regarding the `istanbul` path and version.. I think adding `istanbul@~0.3.20` as a `devDependency` and referencing it directly will probably cause less problems from a version mismatch than accessing via the deep path.

bajtos commented:
> Reading through the implementation, I don't think it's quite as bad as you described. Or at least, you've done it the cleanest way I can think of.

Thank you :)

> I think adding istanbul@~0.3.20 as a devDependency and referencing it directly will probably cause less problems from a version mismatch than accessing via the deep path.

Done - see c3d4f82.

bajtos commented:
@slnode test please

rmg commented:
Unfortunately this can't be verified on CI until either build-framework injects the required flags or this module runs with coverage by default.

Other than that, LGTM.

rmg commented:
Passing the review baton to @bnoordhuis since he's more qualified to review the specifics of the code changes.

bajtos commented:
@bnoordhuis ping

Commit:
Add nyc to dev dependencies, npm script "coverage"

This is used by build-framework as a flag to run the tests via NYC
in order to get code coverage.

The script "npm run coverage" allows module developers to easily
compute the code coverage locally.

bajtos commented:
@rmg I've added one more commit e74bde8376cd0df43c7aa3cbfa123bfae6d57e46 to trigger NYC-detection in build-framework (see https://github.com/strongloop-internal/build-framework/pull/10). Does it still LGTY?

@bnoordhuis ping, could you please review?

Commit:
Ignore code-coverage artifacts

bnoordhuis commented on src/worker-bindings.cc:
```diff
@@ -154,6 +155,48 @@ const char* Worker::InitIsolate() {
+#else
+  Handle<Context> context = context_;
+  Context::Scope context_scope(context_);
+#endif
```
Are you using this logic in more places?  If yes, consider wrapping it in a scoped class.

bajtos replied:
Good catch, I'll do that in a new PR.

bajtos replied:
Well, it turns out I need the `context` variable in the code below (later). What do you propose - should I create something like `ContextWithScope` and then retrieve the context handle via `contextWithScope.context`? Do you have any better suggestion?

```cpp
ContextWithScope ctx(this); // uses this->isolate_ and this->context_
// later
Local<Object> global(ctx.context->Global());

// or perhaps implement -> operator, so that I can write
Local<Object> global(ctx->Global());
```

@bnoordhuis Thoughts?

bnoordhuis replied:
I'd pass in `isolate_` and `context_` explicitly and have a `context` property or getter.  Only go through the trouble if there are at least three places where this is duplicated, else it's not worth the effort.

bajtos replied:
I decided to use a different approach which is IMO simpler and hopefully easier to maintain - see #21. I thought there were three places where this code was duplicated, but as it turns out, there are only two of them :-] I still think the change is worth it.

bnoordhuis commented on src/worker-bindings.cc:
```diff
@@ -154,6 +155,48 @@ const char* Worker::InitIsolate() {
+  key = Nan::New("stringify").ToLocalChecked();
+  Local<Function> stringify = Nan::Get(json, key).ToLocalChecked().As<Function>();
+
+  Local<String> coverageString = stringify->Call(json, 1, &coverage).As<String>();
```
Long lines.

bajtos replied:
Good catch, I'll fix both long lines and casing to snake_case.

bnoordhuis commented:
LGTM, no substantial comments.
